File: fs/Practice2/DynamicProgramming/Bit/Schedule2.fs
// https://atcoder.jp/contests/joi2014yo/tasks/joi2014yo_d

(
  stdin.ReadLine () |> int,
  stdin.ReadLine ()
  |> Seq.map (
    fun c ->
      match c with
      | 'J' -> 1
      | 'O' -> 2
      | 'I' -> 4
      | _   -> 0
  )
  |> Seq.toArray
)
|> fun (n, rs) ->
  [0..n - 1]
  |> List.fold (
    fun (memo : int []) i ->
      [|1..7|]
      |> Array.map (
        fun j ->
          if j &&& rs.[i] <> rs.[i]
          then 0
          else (
            [1..7]
            |> List.map (
              fun k ->
                if [0..2] |> List.exists (
                  fun x -> (j >>> x &&& 1 = 1) && (k >>> x &&& 1 = 1)
                )
                then memo.[k - 1] % 10007
                else 0
            )
            |> List.sum
          )
      )
  ) ([[|1|]; Array.create 6 0] |> Array.concat)
  |> Seq.sum
  |> fun v -> v % 10007
|> printfn "%d"

// ----------

// i in 1..n

// j: IOJ
// 0: 000
// 1: 001
// 2: 010
// 3: 011
// 4: 100
// 5: 101
// 6: 110
// 7: 111

// ----------

// input:
// 2
// OI

// memo:
// [|
//   [|0; 1; 0; 0; 0; 0; 0; 0|]
//   [|0; 0; 0; 1; 0; 0; 0; 1|]
//   [|0; 0; 0; 0; 1; 2; 2; 2|]
// |]

// output:
// 7

// https://atcoder.jp/contests/joi2014yo/submissions/20865427

File: scala/Practice2/DynamicProgramming/Bit/Schedule2.scala
// https://atcoder.jp/contests/joi2014yo/tasks/joi2014yo_d

object Main {

  implicit class Pipeline[T](x: T) {
    def |>[S](f: T => S): S = f(x)
  }

  def main(args: Array[String]) = (
    (
      io.StdIn.readLine.toInt,
      io.StdIn.readLine
        .map(c =>
          c match {
            case 'J' => 1
            case 'O' => 2
            case 'I' => 4
            case _   => 0
          }
        )
        .toArray
    )
    |> { case (n, rs) => (
      (0 until n)
        .foldLeft(
          Array.concat(Array(1), Array.fill[Int](6)(0))
        )((memo, i) =>
          Array.range(1, 8).map(j =>
            if ((j & rs(i)) != rs(i))
              0
            else (
              (1 to 7).map(k =>
                if ((0 to 2).exists(x => ((j >> x & 1) == 1) && ((k >> x & 1) == 1)))
                  memo(k - 1) % 10007
                else
                  0
              )
              .sum
            )
          )
        )
        .sum
        |> (v => v % 10007)
    )}
    |> println
  )
}

// https://atcoder.jp/contests/joi2014yo/submissions/20928844
